From: Ursula Braun <ursula.braun@de.ibm.com>
Subject: qeth: request length checking in snmp ioctl
Patch-mainline: not yet
Git-commit: -
References: bnc#847660, LTC#99511

Description:  qeth: request length checking in snmp ioctl
Symptom:      Kernel crash in case of exploit.
Problem:      Missing check of request length.
Solution:     Check user-defined length in snmp ioctl request and allow request
              only if it fits into a qeth command buffer.
Reproduction: Write user space program invoking the qeth SNMP ioctl with a
              request length greater than 1 page.

Signed-off-by: Ursula Braun <ursula.braun@de.ibm.com>
Acked-by: John Jolly <jjolly@suse.de>
---
 drivers/s390/net/qeth_core_main.c |    6 +++++-
 1 file changed, 5 insertions(+), 1 deletion(-)

--- a/drivers/s390/net/qeth_core_main.c
+++ b/drivers/s390/net/qeth_core_main.c
@@ -4429,7 +4429,7 @@ int qeth_snmp_command(struct qeth_card *
 	struct qeth_cmd_buffer *iob;
 	struct qeth_ipa_cmd *cmd;
 	struct qeth_snmp_ureq *ureq;
-	int req_len;
+	unsigned int req_len;
 	struct qeth_arp_query_info qinfo = {0, };
 	int rc = 0;
 
@@ -4445,6 +4445,10 @@ int qeth_snmp_command(struct qeth_card *
 	/* skip 4 bytes (data_len struct member) to get req_len */
 	if (copy_from_user(&req_len, udata + sizeof(int), sizeof(int)))
 		return -EFAULT;
+	if (req_len > (QETH_BUFSIZE - IPA_PDU_HEADER_SIZE -
+		       sizeof(struct qeth_ipacmd_hdr) -
+		       sizeof(struct qeth_ipacmd_setadpparms_hdr)))
+			return -EINVAL;
 	ureq = memdup_user(udata, req_len + sizeof(struct qeth_snmp_ureq_hdr));
 	if (IS_ERR(ureq)) {
 		QETH_CARD_TEXT(card, 2, "snmpnome");
